fix(lint): make linting work properly again

eventualbuddha opened the pull request:
## Overview
<!-- add a link to a Github Issue here -->
I'm not sure why, but linting has not been working 100% for a little while. It has been working to some degree because our prettier formatting is done through eslint, but some other rules (like `prefer-readonly`) were not being enforced. This PR fixes that, possibly by fixing the `vx/gts-spread-like-types` rule.

We've been seeing false positives with this rule somtimes. Turns out it's a difference in the TypeScript version from when we upgraded from 4.2.x to 4.6.x, but I didn't catch it during the upgrade because I forgot to upgrade the linter package's copy of TypeScript. This fixes it so it should work in both versions, though we only really care about 4.6.x.

## Demo Video or Screenshot
n/a

## Testing Plan 
Automated tests, navigating through various projects in VS Code to ensure the ESLint integration is working well.

## Checklist
- [ ] I have added [logging](https://github.com/votingworks/vxsuite/tree/main/libs/logging) where appropriate to any new user actions, system updates such as file reads or storage writes, or errors introduced.
- [ ] I have added JSDoc comments to any newly introduced exports
<!-- for user-facing changes, non-user facing changes can remove or ignore the below items -->
- [ ] I have added a screenshot and/or video to this PR to demo the change
- [ ] I have added the "user_facing_change" label to this PR to automate an announcement in #machine-product-updates

benadida reviewed (commented):
This is a pretty big change given the addition of all the `useCallback`s right? Should we do some extensive manual testing before merging? Or is there a reason not to be too concerned?

eventualbuddha commented:
> This is a pretty big change given the addition of all the `useCallback`s right? Should we do some extensive manual testing before merging? Or is there a reason not to be too concerned?

For reasons I still don't understand, the `react/jsx-no-bind` errors stopped showing up after I rebased on `main` again. I do think we'll want to get to that in the future, but I removed the fixes for those for now, leaving only the (much safer) `prefer-readonly` fixes.

jonahkagan reviewed (approved):
Nice! The `gts-spread-like-types` rule failures were so confusing. Glad you looked into them.
